--- a/arch/arm64/boot/dts/rockchip/rk3399-nanopi-r4s.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3399-nanopi-r4s.dts
@@ -62,6 +62,65 @@
 		regulator-always-on;
 		regulator-boot-on;
 	};
+
+    fan: pwm-fan {
+        compatible = "pwm-fan";
+        cooling-levels = <0 18 102 170 255>;
+        #cooling-cells = <2>;
+        fan-supply = <&vdd_5v>;
+        pwms = <&pwm1 0 50000 0>;
+    };
+};
+
+/* Trip index starts at 3 */
+&cpu_thermal {
+    trips {
+        cpu_trip1: cpu_trip1 {
+            temperature = <40000>;
+            hysteresis = <2000>;
+            type = "active";
+        };
+
+        cpu_trip2: cpu_trip2 {
+            temperature = <50000>;
+            hysteresis = <2000>;
+            type = "active";
+        };
+
+        cpu_trip3: cpu_trip3 {
+            temperature = <60000>;
+            hysteresis = <2000>;
+            type = "active";
+        };
+
+        cpu_trip4: cpu_trip4 {
+            temperature = <70000>;
+            hysteresis = <2000>;
+            type = "active";
+        };
+    };
+
+	cooling-maps {
+        map1 {
+            trip = <&cpu_trip1>;
+            cooling-device = <&fan 1 1>;
+        };
+
+        map2 {
+            trip = <&cpu_trip2>;
+            cooling-device = <&fan 2 2>;
+        };
+
+        map3 {
+            trip = <&cpu_trip3>;
+            cooling-device = <&fan 3 3>;
+        };
+
+        map4 {
+            trip = <&cpu_trip4>;
+            cooling-device = <&fan 4 THERMAL_NO_LIMIT>;
+        };
+    };
 };
 
 &emmc_phy {
